File: 프로그래머스/lv3/87391. 공 이동 시뮬레이션/공 이동 시뮬레이션.py
def solution(n, m, x, y, queries):
    col_q, row_q = [], []
    for c, d in queries:
        div, mod = divmod(c, 2)
        if div == 0:
            if mod == 0:
                col_q.append(-d)
            else:
                col_q.append(d)
        else:
            if mod == 0:
                row_q.append(-d)
            else:
                row_q.append(d)

    def sub_f(axios_q: list, dest: int, wall: int):
        mn, mx = dest, dest
        while axios_q:
            dd = axios_q.pop()
            if dd < 0:
                if mn > 0:
                    mn = max(0, mn - dd)
                else:
                    if mx - dd < 0:
                        return 0
                mx = min(mx - dd, wall - 1)
            else:
                if mx < wall - 1:
                    mx = min(mx - dd, wall - 1)
                else:
                    if mn - dd > wall - 1:
                        return 0
                mn = max(0, mn - dd)
        return (mx - mn) + 1
    col = sub_f(col_q, y, m)
    row = sub_f(row_q, x, n)
    return col * row

File: 프로그래머스/lv3/87391. 공 이동 시뮬레이션/공 이동 시뮬레이션.java
import java.util.Stack;
import java.util.LinkedList;

class Solution {

    
    public long solution(int n, int m, int x, int y, int[][] queries) {
        long answer = 0;
        Stack[] stacks = new Stack[2];
        stacks[0] = new Stack<Integer>();
        stacks[1] = new Stack<Integer>();
        
        for (int[] q : queries) {
            int div = q[0] / 2;
            int mod = q[0] % 2;
            if (mod == 0) stacks[div].push(-q[1]);
            else stacks[div].push(q[1]);
        }
        return (long) (availRange(stacks[0], y, m)) * (long) (availRange(stacks[1], x, n));
    }
    
    private int availRange(Stack<Integer> q, int dest, int wall) {
        int mn = dest;
        int mx = dest;
        while (!q.isEmpty()) {
            int dd = q.pop();
            if (dd < 0) {
                if (mn > 0) mn = Math.max(0, mn - dd);
                else if (mx - dd < 0) return 0;
                mx = Math.min(mx - dd, wall - 1);
            } else {
                if (mx < wall - 1) mx = Math.min(mx - dd, wall - 1);
                else if (mn - dd > wall - 1) return 0;
                mn = Math.max(0, mn - dd);
            }
        }
        return mx - mn + 1;
    }
}